File: src/app/src/main/java/com/hsousa_apps/Autocarros/data/Instruction.kt
package com.hsousa_apps.Autocarros.data;

import com.google.gson.JsonObject
import org.json.JSONArray
import org.json.JSONObject


class Instruction {
    var routes: MutableList<StepRoute> =  mutableListOf()

    fun init_instructions(routes: JSONArray): Instruction {

        for (i in 0 until routes.length())
            this.routes.add(init_step_route(routes.getJSONObject(i)))

        return this
    }

    fun init_step_route(json: JSONObject): StepRoute {
        var instruction_route = StepRoute()

        instruction_route.overview_polyline_points = json.getJSONObject("overview_polyline").getString("points")
        instruction_route.warnings = json.getJSONArray("warnings")
        var legs = json.getJSONArray("legs")
        for (j in 0 until legs.length())
            instruction_route.legs.add(init_leg(legs.getJSONObject(j)))

        return instruction_route
    }

    fun init_leg(json: JSONObject): Leg {
        var leg = Leg()

        leg.start_address = json.getString("start_address")
        val start_location = json.getJSONObject("start_location")
        leg.start_location = Location(start_location.getDouble("lat"), start_location.getDouble("lng"))
        leg.end_address = json.getString("end_address")
        val end_location = json.getJSONObject("end_location")
        leg.end_location = Location(end_location.getDouble("lat"), end_location.getDouble("lng"))

        leg.departure = json.getJSONObject("departure_time").getString("text")
        leg.arrival = json.getJSONObject("arrival_time").getString("text")
        leg.duration = json.getJSONObject("duration").getString("text")

        val steps = json.getJSONArray("steps")
        for (k in 0 until steps.length())
            leg.steps.add(init_step(steps.getJSONObject(k)))

        return leg
    }

    fun init_step(json: JSONObject): Step {
        var step = Step()

        step.instructions = json.getString("html_instructions")

        step.start_location = Location(json.getJSONObject("start_location").getDouble("lat"), json.getJSONObject("start_location").getDouble("lng"))
        step.end_location = Location(json.getJSONObject("end_location").getDouble("lat"), json.getJSONObject("end_location").getDouble("lng"))


        step.distance = json.getJSONObject("distance").getString("text")
        step.duration = json.getJSONObject("duration").getString("text")

        step.polyline = json.getJSONObject("polyline").getString("points")
        step.travel_mode = json.getString("travel_mode")

        if (json.has("steps")) for (l in 0 until json.getJSONArray("steps").length()) step.steps.add(init_step(json.getJSONArray("steps").getJSONObject(l)))
        else if (json.has("transit_details")) step.transit_details = init_transit_details(json.getJSONObject("transit_details"))

        return step
    }

    fun init_transit_details(json: JSONObject) : TransitDetails{
        var transit_details = TransitDetails()


        transit_details.departure_stop = json.getJSONObject("departure_stop").getString("name")
        transit_details.departure_location = Location(
            json.getJSONObject("departure_stop").getJSONObject("location").getDouble("lat"),
            json.getJSONObject("departure_stop").getJSONObject("location").getDouble("lng"),
        )
        transit_details.departure_time = json.getJSONObject("departure_time").getString("text")


        transit_details.arrival_stop = json.getJSONObject("arrival_stop").getString("name")
        transit_details.arrival_location = Location(
            json.getJSONObject("arrival_stop").getJSONObject("location").getDouble("lat"),
            json.getJSONObject("arrival_stop").getJSONObject("location").getDouble("lng"),
        )
        transit_details.arrival_time = json.getJSONObject("arrival_time").getString("text")

        transit_details.headsign = json.getString("headsign")
        transit_details.line = init_line(json.getJSONObject("line"))
        transit_details.num_stops = json.getInt("num_stops")

        return transit_details
    }

    fun init_line(json: JSONObject): Line{
        var line = Line()
        var agencies = mutableListOf<Agency>()

        var agencies_json = json.getJSONArray("agencies")
        for (i in 0 until agencies_json.length())
            agencies.add(init_agency(agencies_json.getJSONObject(i)))

        line.name = json.getString("name")
        line.short_name = json.getString("short_name")

        line.vehicle = init_vehicle(json.getJSONObject("vehicle"))

        return line
    }

    fun init_agency(json: JSONObject): Agency {
        var agency = Agency()

        agency.name = json.getString("name")
        agency.phone = json.getString("phone")
        agency.url = json.getString("url")

        return agency
    }

    fun init_vehicle(json: JSONObject): Vehicle {
        var vehicle = Vehicle()

        vehicle.icon = json.getString("icon")
        vehicle.name = json.getString("name")
        vehicle.type = json.getString("type")

        return vehicle
    }
    override fun toString(): String {
        val json = JSONObject()
        val routesArray = JSONArray()

        for (route in routes) {
            val routeObject = JSONObject()
            val legsArray = JSONArray()

            for (leg in route.legs) {
                val legObject = JSONObject()

                legObject.put("start_address", leg.start_address)
                legObject.put("end_address", leg.end_address)
                legObject.put("duration", leg.duration)
                legObject.put("departure", leg.departure)
                legObject.put("arrival", leg.arrival)

                val stepsArray = JSONArray()

                for (step in leg.steps) {
                    val stepObject = JSONObject()

                    stepObject.put("instructions", step.instructions)
                    stepObject.put("travel_mode", step.travel_mode)
                    stepObject.put("distance", step.distance)
                    stepObject.put("duration", step.duration)
                    stepObject.put("polyline", step.polyline)

                    if (step.travel_mode == "TRANSIT") {
                        val transitDetailsObject = JSONObject()

                        transitDetailsObject.put("departure_stop", step.transit_details.departure_stop)
                        transitDetailsObject.put("departure_location", JSONObject()
                            .put("lat", step.transit_details.departure_location.x)
                            .put("lng", step.transit_details.departure_location.y)
                        )
                        transitDetailsObject.put("departure_time", step.transit_details.departure_time)

                        transitDetailsObject.put("arrival_stop", step.transit_details.arrival_stop)
                        transitDetailsObject.put("arrival_location", JSONObject()
                            .put("lat", step.transit_details.arrival_location.x)
                            .put("lng", step.transit_details.arrival_location.y)
                        )
                        transitDetailsObject.put("arrival_time", step.transit_details.arrival_time)

                        transitDetailsObject.put("headsign", step.transit_details.headsign)

                        val lineObject = JSONObject()

                        lineObject.put("name", step.transit_details.line.name)
                        lineObject.put("short_name", step.transit_details.line.short_name)

                        val vehicleObject = JSONObject()

                        vehicleObject.put("icon", step.transit_details.line.vehicle.icon)
                        vehicleObject.put("name", step.transit_details.line.vehicle.name)
                        vehicleObject.put("type", step.transit_details.line.vehicle.type)

                        lineObject.put("vehicle", vehicleObject)

                        val agenciesArray = JSONArray()

                        for (agency in step.transit_details.line.agencies) {
                            val agencyObject = JSONObject()

                            agencyObject.put("name", agency.name)
                            agencyObject.put("phone", agency.phone)
                            agencyObject.put("url", agency.url)

                            agenciesArray.put(agencyObject)
                        }

                        lineObject.put("agencies", agenciesArray)

                        transitDetailsObject.put("line", lineObject)
                        transitDetailsObject.put("num_stops", step.transit_details.num_stops)

                        stepObject.put("transit_details", transitDetailsObject)
                    } else if (step.travel_mode == "WALKING"){
                        val stepStepsArray = JSONArray()

                        for (stepsStep in step.steps){
                            val stepStepsStepObject = JSONObject()

                            stepStepsStepObject.put("instructions", stepsStep.instructions)
                            stepStepsStepObject.put("travel_mode", stepsStep.travel_mode)
                            stepStepsStepObject.put("distance", stepsStep.distance)
                            stepStepsStepObject.put("duration", stepsStep.duration)
                            stepStepsStepObject.put("polyline", stepsStep.polyline)

                            stepStepsArray.put(stepStepsStepObject)
                        }
                    }

                    stepsArray.put(stepObject)
                }

                legObject.put("steps", stepsArray)
                legsArray.put(legObject)
            }

            routeObject.put("legs", legsArray)
            routeObject.put("overview_polyline_points", route.overview_polyline_points)
            routeObject.put("warnings", route.warnings)

            routesArray.put(routeObject)
        }

        json.put("routes", routesArray)

        return json.toString(4)
    }
}

class StepRoute {
    var legs: MutableList<Leg> = mutableListOf()
    lateinit var overview_polyline_points: String
    lateinit var warnings: JSONArray
}

class Leg {
    lateinit var start_address: String
    lateinit var start_location: Location
    lateinit var end_address: String
    lateinit var end_location: Location

    lateinit var departure: String
    lateinit var arrival: String
    lateinit var duration: String

    var steps: MutableList<Step> = mutableListOf()
}

class Step {
    lateinit var instructions: String

    lateinit var start_location: Location
    lateinit var end_location: Location

    lateinit var maneuver: String
    lateinit var transit_details: TransitDetails

    lateinit var distance: String
    lateinit var duration: String

    lateinit var travel_mode: String
    lateinit var polyline: String

    var steps: MutableList<Step> = mutableListOf()
}

class TransitDetails {
    lateinit var departure_location: Location
    lateinit var departure_stop: String
    lateinit var departure_time: String
    lateinit var arrival_location: Location
    lateinit var arrival_stop: String
    lateinit var arrival_time: String

    lateinit var headsign: String
    lateinit var line: Line
    var num_stops: Int = -1
}

class Line {
    var agencies: MutableList<Agency> = mutableListOf()

    lateinit var name: String
    lateinit var short_name: String

    lateinit var vehicle: Vehicle
}

class Agency {
    lateinit var name: String
    lateinit var phone: String
    lateinit var url: String
}

class Vehicle {
    lateinit var icon: String
    lateinit var name: String
    lateinit var type: String
}

File: src/saomiguelbus/lib/models/instruction.dart
import 'dart:convert';

import 'package:saomiguelbus/models/index.dart';

class Instruction {
  List<StepRoute> routes = [];

  Instruction initInstructions(List<dynamic> routesJson) {
    for (var routeJson in routesJson) {
      routes.add(initStepRoute(routeJson));
    }
    return this;
  }

  StepRoute initStepRoute(dynamic json) {
    var instructionRoute = StepRoute();

    instructionRoute.overviewPolylinePoints =
        json['overview_polyline']['points'];
    instructionRoute.warnings = List<String>.from(json['warnings']);
    var legs = json['legs'];
    for (var legJson in legs) {
      instructionRoute.legs.add(initLeg(legJson));
    }

    return instructionRoute;
  }

  Leg initLeg(dynamic json) {
    var leg = Leg();

    leg.startAddress = json['start_address'];
    leg.startLocation =
        Location(json['start_location']['lat'], json['start_location']['lng']);
    leg.endAddress = json['end_address'];
    leg.endLocation =
        Location(json['end_location']['lat'], json['end_location']['lng']);

    leg.departure = json['departure_time']['text'];
    leg.arrival = json['arrival_time']['text'];
    leg.duration = json['duration']['text'];

    var steps = json['steps'];
    for (var stepJson in steps) {
      leg.steps.add(initStep(stepJson, leg));
    }

    return leg;
  }

  Step initStep(dynamic json, Leg? leg) {
    var step = Step();
    step.leg = leg;

    try {
      step.instructions = json['html_instructions'];
    } catch (e) {
      step.instructions = '';
    }

    step.startLocation =
        Location(json['start_location']['lat'], json['start_location']['lng']);
    step.endLocation =
        Location(json['end_location']['lat'], json['end_location']['lng']);

    step.distance = json['distance']['text'];
    step.duration = json['duration']['text'];

    step.polyline = json['polyline']['points'];
    step.travelMode = json['travel_mode'];

    if (json.containsKey('steps')) {
      for (var stepJson in json['steps']) {
        step.steps.add(initStep(stepJson, null));
      }
    } else if (json.containsKey('transit_details')) {
      step.transitDetails = initTransitDetails(json['transit_details']);
    }

    return step;
  }

  TransitDetails initTransitDetails(dynamic json) {
    var transitDetails = TransitDetails();

    transitDetails.departureStop = json['departure_stop']['name'];
    transitDetails.departureLocation = Location(
        json['departure_stop']['location']['lat'],
        json['departure_stop']['location']['lng']);
    transitDetails.departureTime = json['departure_time']['text'];

    transitDetails.arrivalStop = json['arrival_stop']['name'];
    transitDetails.arrivalLocation = Location(
        json['arrival_stop']['location']['lat'],
        json['arrival_stop']['location']['lng']);
    transitDetails.arrivalTime = json['arrival_time']['text'];

    transitDetails.headsign = json['headsign'];
    transitDetails.line = initLine(json['line']);
    transitDetails.numStops = json['num_stops'];

    return transitDetails;
  }

  Line initLine(dynamic json) {
    var line = Line();
    var agencies = <Agency>[];

    var agenciesJson = json['agencies'];
    for (var agencyJson in agenciesJson) {
      agencies.add(initAgency(agencyJson));
    }

    line.agencies = agencies;
    line.name = json['name'];
    line.shortName = json['short_name'];

    line.vehicle = initVehicle(json['vehicle']);

    return line;
  }

  Agency initAgency(dynamic json) {
    var agency = Agency();

    agency.name = json['name'];
    agency.phone = json['phone'];
    agency.url = json['url'];

    return agency;
  }

  Vehicle initVehicle(dynamic json) {
    var vehicle = Vehicle();

    vehicle.icon = json['icon'];
    vehicle.name = json['name'];
    vehicle.type = json['type'];

    return vehicle;
  }

  @override
  String toString() {
    var json = <String, dynamic>{};
    var routesArray = <dynamic>[];

    for (var route in routes) {
      var routeObject = <String, dynamic>{};
      var legsArray = <dynamic>[];

      for (var leg in route.legs) {
        var legObject = <String, dynamic>{};

        legObject['start_address'] = leg.startAddress;
        legObject['end_address'] = leg.endAddress;
        legObject['duration'] = leg.duration;
        legObject['departure'] = leg.departure;
        legObject['arrival'] = leg.arrival;

        var stepsArray = <dynamic>[];

        for (var step in leg.steps) {
          var stepObject = <String, dynamic>{};

          stepObject['instructions'] = step.instructions;
          stepObject['travel_mode'] = step.travelMode;
          stepObject['distance'] = step.distance;
          stepObject['duration'] = step.duration;
          stepObject['polyline'] = step.polyline;

          if (step.travelMode == 'TRANSIT') {
            var transitDetailsObject = <String, dynamic>{};

            transitDetailsObject['departure_stop'] =
                step.transitDetails.departureStop;
            transitDetailsObject['departure_location'] = {
              'lat': step.transitDetails.departureLocation.latitude,
              'lng': step.transitDetails.departureLocation.longitude,
            };
            transitDetailsObject['departure_time'] =
                step.transitDetails.departureTime;

            transitDetailsObject['arrival_stop'] =
                step.transitDetails.arrivalStop;
            transitDetailsObject['arrival_location'] = {
              'lat': step.transitDetails.arrivalLocation.latitude,
              'lng': step.transitDetails.arrivalLocation.longitude,
            };
            transitDetailsObject['arrival_time'] =
                step.transitDetails.arrivalTime;

            transitDetailsObject['headsign'] = step.transitDetails.headsign;

            var lineObject = <String, dynamic>{};

            lineObject['name'] = step.transitDetails.line.name;
            lineObject['short_name'] = step.transitDetails.line.shortName;

            var vehicleObject = <String, dynamic>{};

            vehicleObject['icon'] = step.transitDetails.line.vehicle.icon;
            vehicleObject['name'] = step.transitDetails.line.vehicle.name;
            vehicleObject['type'] = step.transitDetails.line.vehicle.type;

            lineObject['vehicle'] = vehicleObject;

            var agenciesArray = <dynamic>[];

            for (var agency in step.transitDetails.line.agencies) {
              var agencyObject = <String, dynamic>{};

              agencyObject['name'] = agency.name;
              agencyObject['phone'] = agency.phone;
              agencyObject['url'] = agency.url;

              agenciesArray.add(agencyObject);
            }

            lineObject['agencies'] = agenciesArray;

            transitDetailsObject['line'] = lineObject;
            transitDetailsObject['num_stops'] = step.transitDetails.numStops;

            stepObject['transit_details'] = transitDetailsObject;
          } else if (step.travelMode == 'WALKING') {
            var stepStepsArray = <dynamic>[];

            for (var stepsStep in step.steps) {
              var stepStepsStepObject = <String, dynamic>{};

              stepStepsStepObject['instructions'] = stepsStep.instructions;
              stepStepsStepObject['travel_mode'] = stepsStep.travelMode;
              stepStepsStepObject['distance'] = stepsStep.distance;
              stepStepsStepObject['duration'] = stepsStep.duration;
              stepStepsStepObject['polyline'] = stepsStep.polyline;

              stepStepsArray.add(stepStepsStepObject);
            }
          }

          stepsArray.add(stepObject);
        }

        legObject['steps'] = stepsArray;
        legsArray.add(legObject);
      }

      routeObject['legs'] = legsArray;
      routeObject['overview_polyline_points'] = route.overviewPolylinePoints;
      routeObject['warnings'] = route.warnings;

      routesArray.add(routeObject);
    }

    json['routes'] = routesArray;

    return jsonEncode(json);
  }
}

class StepRoute {
  List<Leg> legs = [];
  late String overviewPolylinePoints;
  late List<dynamic> warnings;
}

class Leg {
  late String startAddress;
  late Location startLocation;
  late String endAddress;
  late Location endLocation;

  late String departure;
  late String arrival;
  late String duration;

  List<Step> steps = [];
}

class Step {
  Leg? leg;
  late String instructions;

  late Location startLocation;
  late Location endLocation;

  late String maneuver;
  late TransitDetails transitDetails;

  late String distance;
  late String duration;

  late String travelMode;
  late String polyline;

  List<Step> steps = [];
}

class TransitDetails {
  late Location departureLocation;
  late String departureStop;
  late String departureTime;
  late Location arrivalLocation;
  late String arrivalStop;
  late String arrivalTime;

  late String headsign;
  late Line line;
  int numStops = -1;
}

class Line {
  List<Agency> agencies = [];

  late String name;
  late String shortName;

  late Vehicle vehicle;
}

class Agency {
  late String name;
  late String phone;
  late String url;
}

class Vehicle {
  late String icon;
  late String name;
  late String type;
}
